### PR TITLE
distance based me

### PR DESCRIPTION
**Pull request by okhlif**
### Description
_________________________________________________________________________________

- Distance based ME search region instead of fixed search region
### Author(s)
_________________________________________________________________________________
@okhlif 
@hguermaz 

### Type of change
_________________________________________________________________________________
Enhancement.
### Tests and performance
_________________________________________________________________________________
Data generated on the full objective-1-fast data set in context in enc-mode 0.

Speed Deviation | BD-Rate Deviation
-- | --
tbd% | -0.33%

### CODE REVIEW

File: Source/Lib/Common/Codec/EbDefinitions.h [hunk 1]
```diff
@@ -50,6 +50,7 @@ extern "C" {
 #define SWITCHED_HALF_PEL_MODE   0 //MUlti-Stage ME - adapt half_pel mode based on fulpel distortion
 #endif
 #define MC_DYNAMIC_PAD              1
+#define DIST_BASED_ME_SEARCH_AREA       1 // Distance based multiped for ME search region
```
Comment (bzhutencent) [inline Source/Lib/Common/Codec/EbDefinitions.h]: what does multiped mean?

File: Source/Lib/Common/Codec/EbDefinitions.h [hunk 2]
```diff
@@ -3374,7 +3375,140 @@ static const uint16_t hme_level2_search_area_in_height_array_bottom[SC_MAX_LEVEL
...
+    },{
+    // SC
+        { 480,    480,     480,     480,     368,    368,    280,    280,    280,    280,    280,    280,    280 },
+        { 480,    480,     480,     480,     368,    368,    280,    280,    280,    280,    280,    280,    280 },
+        { 1024,   1024,    1024,    1024,    784,    784,    600,    600,    600,    600,    600,    600,    600 },
+        { 1024,   1024,    1024,    1024,    784,    784,    600,    600,    600,    600,    600,    600,    600 }
+    }
+};
+    //    M0      M1      M2      M3   M4     M5     M6     M7     M8     M9     M10    M11    M12
+static const uint16_t min_search_area_width[SC_MAX_LEVEL][INPUT_SIZE_COUNT][MAX_SUPPORTED_MODES] = {
```
Comment (bzhutencent) [inline Source/Lib/Common/Codec/EbDefinitions.h]: max is named with max_me_search, can we rename to min_me_search_width to be consistent

File: Source/Lib/Common/Codec/EbDefinitions.h [hunk 3]
```diff
@@ -3374,7 +3375,140 @@ static const uint16_t hme_level2_search_area_in_height_array_bottom[SC_MAX_LEVEL
...
+    } , {
+    // SC
+        { 60,    60,    60,    60,    48,    48,    40,    40,    40,    40,    40,    40,    40 },
+        { 60,    60,    60,    60,    48,    48,    40,    40,    40,    40,    40,    40,    40 },
+        { 128,   128,   128,   128,   96,    96,    72,    72,    72,    72,    72,    72,    72 },
+        { 128,   128,   128,   128,   96,    96,    72,    72,    72,    72,    72,    72,    72 }
+    }
+};
+    //    M0      M1      M2      M3   M4     M5     M6    M7     M8     M9     M10   M11    M12
+static const uint16_t min_tf_search_area_width[SC_MAX_LEVEL][INPUT_SIZE_COUNT][MAX_SUPPORTED_MODES] = {
```
Comment (bzhutencent) [inline Source/Lib/Common/Codec/EbDefinitions.h]: same here, can the name be consistent with max?

File: Source/Lib/Common/Codec/EbDefinitions.h (hunk 1 above)
Comment (okhlif) [reply Source/Lib/Common/Codec/EbDefinitions.h]: typo, multiplier  I meant. I will fix it thanks

File: Source/Lib/Encoder/Codec/EbMotionEstimation.c [hunk 4]
```diff
@@ -9777,6 +9777,58 @@ void integer_search_sb(
...
+            // Update ME search reagion size based on hme-data
+#if SC_HME_PRUNING
+            if (context_ptr->reduce_me_sr_flag[list_index][ref_pic_index] == SC_HME_TH_STILL) {
+                search_area_width = ((search_area_width / SC_SR_DENOM_STILL) + 7) & ~0x07;
+                search_area_height = (search_area_height / SC_SR_DENOM_STILL);
+            }
+            else if (context_ptr->reduce_me_sr_flag[list_index][ref_pic_index] == SC_HME_TH_EASY) {
+                search_area_width = ((search_area_width / SC_SR_DENOM_EASY) + 7) & ~0x07;
+                search_area_height = (search_area_height / SC_SR_DENOM_EASY);
+            }else if (context_ptr->reduce_me_sr_flag[list_index][ref_pic_index]) {
```
Comment (bzhutencent) [inline Source/Lib/Encoder/Codec/EbMotionEstimation.c]: minor: coding style, we either use } elseif , or start a new line with else if

File: Source/Lib/Common/Codec/EbDefinitions.h (hunk 3 above)
Comment (okhlif) [reply Source/Lib/Common/Codec/EbDefinitions.h]: Yes, Thank you

File: Source/Lib/Common/Codec/EbDefinitions.h (hunk 2 above)
Comment (okhlif) [reply Source/Lib/Common/Codec/EbDefinitions.h]: Yes, Thank you

File: Source/Lib/Common/Codec/EbDefinitions.h (hunk 1 above)
Comment (fouedb) [inline Source/Lib/Common/Codec/EbDefinitions.h]: Use "Distance-based"

File: Source/Lib/Encoder/Codec/EbMotionEstimation.c [hunk 5]
```diff
@@ -9777,6 +9777,58 @@ void integer_search_sb(
...
             y_search_center = context_ptr->hme_results[list_index][ref_pic_index].hme_sc_y;
+#if DIST_BASED_ME_SEARCH_AREA
+            search_area_width = context_ptr->search_area_width;
+            search_area_height = context_ptr->search_area_height;
+
+            uint16_t dist = (context_ptr->me_alt_ref == EB_TRUE) ?
+                ABS((int16_t)(context_ptr->tf_frame_index - context_ptr->tf_index_center)) :
+                ABS((int16_t)(pcs_ptr->picture_number -
+                    pcs_ptr->ref_pic_poc_array[list_index][ref_pic_index]));
+            // factor to slow down the ME search region growth to MAX
```
Comment (fouedb) [inline Source/Lib/Encoder/Codec/EbMotionEstimation.c]: Fix "slow down""

Comment (okhlif) [reply Source/Lib/Encoder/Codec/EbMotionEstimation.c]: fixed Thanks

File: Source/Lib/Common/Codec/EbDefinitions.h (hunk 1 above)
Comment (okhlif) [reply Source/Lib/Common/Codec/EbDefinitions.h]: Fixed Thanks

File: Source/Lib/Encoder/Codec/EbMotionEstimation.c (hunk 4 above)
Comment (okhlif) [reply Source/Lib/Encoder/Codec/EbMotionEstimation.c]: Fixed Thanks

**Comment by hassount** [comment]
There seems to be a unit test that's failing in this PR, master is passing, #1069 is passing, but this PR is crashing. I am re-running the test to make sure this is not due to the test system

**Comment by okhlif** [comment]
> There seems to be a unit test that's failing in this PR, master is passing, #1069 is passing, but this PR is crashing. I am re-running the test to make sure this is not due to the test system

I just rebased. I did not see the problem again. Let's see after the rebase

**Pull request merged**